pyomo/common/factory.py

```diff
@@ -55,13 +68,15 @@ def fn(cls):
             return cls
         return fn
 
-    #
-    # Support "with" statements. Forgetting to call deactivate
-    # on Plugins is a common source of memory leaks
-    #
-    def __enter__(self):
-        return self
 
-    def __exit__(self, t, v, traceback):
-        pass
+class Factory(CachedFactory):
```

Comment:
Object inheritance is a specialization operation.  If we had Factory->CachedFactory, then the "is a" doesn't completely describe that relationship.  Rather, the CachedFactory would be a Factory with caching enabled.  By contrast, the current design has the semantics that a Factory "is a" CachedFactory with caching disabled.

I think that this hierarchy makes it easier to manage caching.  You can inject it and then control it with a flag.  (Of course, I could have added the caching logic to Factory ... but then we would be having the conversation about why Factory has this logic which is, by default, disabled.)

Reply:
First, I am realizing that "CachedFactory" is a bad name.  A better name may be "`SingletonFactory`".

I don't think the Factory should have *any* logic around caching.  The CachedFactory (Singleton) provides all the functionality of a Factory, *plus* caching.  I think a more natual logic should be:
```
class Factory(object):
    def __init__(self, description=None):
        self._description = description
        self._cls = {}
        self._doc = {}

    def __call__(self, name, **kwds):
        exception = kwds.pop('exception', False)
        name = str(name)
        if not name in self._cls:
            if not exception:
                return None
            if self._description is None:
                raise ValueError("Unknown factory object type: '%s'" % name)
            raise ValueError("Unknown %s: '%s'" % (self._description, name))
        return self._cls[name](**kwds)

class SingletonFactory(Factory):
    def __init__(self, description=None):
        super(SingletonFactory, self).__init__(description)
        self._cache = {}

    def __call__(self, name, **kwds):
        name = str(name)
        if name in not in self._cache:
            self._cache[name] = super(SingletonFactory, self).__call__(name, **kwds)
        return self._cache[name]
```

Reply:
@jsiirola I don't see why Singleton is preferable to Cached.  This is a factory that creates objects and keeps a copy of the object in a ... cache.

The motivation for integrating this logic into a single class, is that the developer may override the __call__ method.  Such an action would break the nested calls to __call__() that you've defined in your hierarchy.  A simpler approach is to have the logic together in the base class, which the user can manage explicitly with a flag.

Note that I've moved this code to PyUtilib.  I've reordered the class hierarchy, though with a simple edit of the caching flag.

Reply:
Cache refers to the implementation, whereas Singleton refers to the behavior (the CachedFactory will only ever produce a single instance of each registered class... barring pickling or deepcopying).  As the user interacts with the class name, IMHO the class name should reflect the behavior  and not the implementation.

Your argument is that someone deriving from CachedData wouldn't call the base class' `__call__` method?  That holds regardless of how the base class or class hierarchy was implemented...

Reply:
Cached seems like a reasonable description of the behavior to me.  But that's my Computer Science roots showing.

Yes, a derived class wouldn't call the base class's method.  And in fact I do that in several places through class inheritance.  Hence, what you propose is a more complex class hierarchy that is fragile when that happens.